### DISCUSSION TITLE
combine, reorg, and rename std.meta and typecons

### DISCUSSION THREAD

**Discussion by crazymonkyyy** [dlang/PhobosV3-Design#24] [Ideas]
Theres a few very important functions here and allot of ... templates of questionable value, lots of naming complaints, etc.

Id suggest a simpler std.meta with: 
Seq(shortened aliasSeq)
tuple
isTuple
toTuple(allows single values to use `foo[0]`, leaves tuples as is) 
nullable (with `opIndex(typeof(null))`, permissive)
optional (with `get(){assert( ! isnull); return value;}`, strict)

etc. cut off near staticMap?

if `Erase` `templateNot` `whiteHole` airnt deleted should be put elsewhere

**Comment by jmdavis** [comment]
std.meta and std.typecons provide completely different functionality from one another. std.meta provides templates for operating on `AliasSeq`s (in some respects, making it miniature version of std.algorithm but for `AliasSeq`s rather than ranges), whereas std.typecons is for "type constructors" and hosts a variety of wrapper types for various purposes.

While altering the exact set of templates provided by std.meta likely makes sense (and some of them may need some reworking), I don't see how it would make sense to combine it with anything from std.typecons, because their purposes are completely different. std.meta also isn't very large, so unless we add a bunch of stuff to it (which I doubt that we will), it doesn't really make sense to break it up or rearrange it either.

Now, because std.typecons is such a hodgepodge of stuff, there's definitely an argument for breaking it up / rearranging it in some fashion, but what exactly makes sense with that depends largely on what we keep. Certainly, some of it should be redesigned on some level, and some of it probably doesn't need to end up in Phobos v3, but each item in there will need to be examined and evaluated individually.

As for `AliasSeq`, we argued to death about what to rename it to from `TypeTuple` years ago, and `AliasSeq` was the result. It's not a great name (none of the suggestions were), but IMHO, it makes no sense to revisit that at this point. It's just bikeshedding where there will be no agreement, because there really isn't a good name for what an `AliasSeq` is, and renaming it would just make it so that everyone would have to learn another name for it - again - rather than simply learning where it got put in Phobos v3.

**Reply by crazymonkyyy** [reply]
> std.meta and std.typecons provide completely different functionality

The most important template from meta is seq, the most important template in typecons is tuple
tuple is a seq in struct; typecons being over complicated hides that fact but it could be simplified down to
```d
alias seq(T...)=T;
auto tuple(T...)(T args){
	struct Tuple{
		enum istuple=true;
		T me; alias me this;
	}
	return Tuple(args);
}
unittest{
	auto foo=tuple(1,"hi");
	assert(foo[0]==1);
	assert(foo[1]=="hi");
}
```
(right now the same logic is "expand" in phoboes, which massively hides its true nature)
